python/lsst/daf/persistence/registries.py

```diff
@@ -339,23 +344,22 @@ def lookup(self, lookupProperties, reference, dataId, **kwargs):
         If key is a 2-item iterable then will look for elements in the repository where the value is between
         the values of key[0] and key[1].
         :param reference: other data types that may be used to search for values.
-        :param **kwargs: nothing needed for sqlite lookup
+        :param checkColumns: if True, check that keys are actually in the registry and ignore them if not
         :return: a list of values that match keys in lookupProperties.
         """
-        if not self.conn:
-            return None
-
-        # input variable sanitization:
-        reference = sequencify(reference)
-        lookupProperties = sequencify(lookupProperties)
-
         cmd = "SELECT DISTINCT "
         cmd += ", ".join(lookupProperties)
         cmd += " FROM " + " NATURAL JOIN ".join(reference)
         valueList = []
         if dataId is not None and len(dataId) > 0:
             whereList = []
             for k, v in dataId.items():
+                if checkColumns:        # check if k is in registry
+                    try:
+                        self.conn.cursor().execute(f"select count({k}) from {reference[0]}")
```

Comment:
This is potentially very slow, and I don't think it's guaranteed that `reference[0]` holds all columns of interest.

As a first step, `SELECT {k} FROM {reference[0]} LIMIT 1` is likely to be much more efficient.

As a next step, `SELECT * FROM {reference[0]} LIMIT 1` and inspecting the returned row's schema would be even better.

Finally, the `NATURAL JOIN`-based `FROM` clause could be used to look at all available tables.

Reply:
I suppose I was assuming an index on the table so the `count(*)` should be fast, but I'm happy to use the `LIMIT 1`; and I was just being lazy with the `JOIN`.
However, inspecting the schema seems un-necessary (and more work).  What is gained over `f'SELECT {k} FROM {" NATURAL JOIN ".join(reference)} LIMIT 1'`?

Reply:
Issuing one query for `SELECT *` and seeing the returned columns is going to be much more efficient than `SELECT {k}` for each column as long as the number of columns is much greater than 1.  In this case, though, I guess the columns are not all the ones in the database but instead all the ones in the `dataId`.  So maybe you really want `SELECT {", ".join(dataId.keys())} FROM ...`

Reply:
Ah, reading still closer, you want individual column-based notification of presence.  So that gets you back to `SELECT *` if the `dataId` contains a lot of columns or your current code if it only has a few.

Reply:
I think we'll usually have only a few fields, and as I need to know which one caused the problem I think I'll stick with what I have.  As we note below, this code only fires when there are fields in the dataId that confuse the registry query, which is moderately uncommon so a little inefficiency is probably OK.  The use case that forced me to create this patch was specifying an int to read a particular HDU when the dataId was unambiguous but not complete.

Reply:
Oh, and I accepted the don't-use-count, use limit, and ask for the join suggestions.